rfcs/0148-pipe-operator.md

```diff
@@ -0,0 +1,330 @@
+---
```

Comment:
(Attempting to corral feedback into a thread.)

@illustris

> ```nix
> nix-repl> double = x: x*2
> 
> # current syntax
> nix-repl> 3 |> (x: 1 + x) |> double |> toString
> "8"
> 
> # proposed alternative
> nix-repl> 3 |> x: 1 + x |> double |> toString
> 
> # current syntax
> nix-repl> 3 |> (x: 1 + (x |> double)) |> toString
> "7"
> 
> # proposed alternative
> nix-repl> 3 |> x: 1 + (x |> double) |> toString
> ```

Both of those proposals seem, to me, to be ‘more likely’ to mean something else:

```nix
3 |> (x: (1 + x |> double |> toString))
3 |> (x: (1 + (x |> double) |> toString))
```

In other words, if `|>` binds more loosely than the lambda-forming `:`, it would be the first thing in the language to do so and thus would be very surprising to me. So far in the language, a lambda always extends as far rightward as it can.

Reply:
### Partially applied pipes

To me this smells like function composition with extra steps. Function composition is discussed in the RFC as well, and while I am not completely opposed to it, my fear is that it reduces code readability given a lack of type annotations. In `x = |> inc |> toString` it is a lot less intuitively clear that `x` is a function.

This is one of these questions where I'd like to see more usage to find out whether this is a thing that comes up sufficiently often in practice that it is worth dealing with.

(Side note: I think we would get this feature for free if Nix had operator sections)

### Lambdas inside pipe

While I agree that not requiring parentheses here would be nice, especially given that one goal of the pipe operator is to reduce parentheses, I don't see any way this would be realistically implementable without throwing the entire language under the bus. Making the binding strength of abstractions context dependent just doesn't sound great

Reply:
> Making the binding strength of abstractions context dependent just doesn't sound great

agreed, it’s a recipe for disaster

Reply:
### Lambdas inside pipe

> Making the binding strength of abstractions context dependent

Right. I didn't think that through. This is a bad idea.

### Partially applied pipes

> In `x = |> inc |> toString` it is a lot less intuitively clear that `x` is a function

In my opinion it is slightly more clear than, for example,
```
concatLines = concatStringsSep "\n"
getBin = getOutput "bin"
```
or any other partially applied function. The open `|>` (or `|`) at the start suggests to me that this is a function.